### DISCUSSION TITLE
Public Key Invalid...But It's Not

### DISCUSSION THREAD

**Discussion by danielbirch** [emailjs-com/help#61] [Q&A] [answered]
Hi there,

I setup EmailJS on my site many months ago, some time ago it stopped working. I compared my code and found that the EmailJS code has changed. I went through the docs again and updated my code. However I'm getting this error:

```
 POST https://api.emailjs.com/api/v1.0/email/send-form 400 (Bad Request)
(index):54 Failed t {status: 400, text: 'The Public Key is invalid. To find this ID, visit https://dashboard.emailjs.com/admin/account'}
``` 

BUT, the public key is 100% correct. I tried also regenerating a new public key and using that, triple checked everything and still no go.

Any help is greatly appreciated. Any ideas where I can look next or what may be going on?

**Comment by xr0master** [comment]
You should look in the Network tab to check what data you are actually sending to the service.
If you are not using SDK, most likely the context type is not correct.

**Reply by danielbirch** [reply]
I am using the SDK using the browser script. When you say to check the network tab, where specifically are you referring to? Is this it (see attached screenshot).
<img width="1470" alt="Screenshot 2023-10-22 at 10 35 19 AM" src="https://github.com/emailjs-com/help/assets/122473191/665695c6-bb06-487d-8bd2-5c5ab4d996a8">

I'm just not sure how this related to the error of the public key being wrong (even though it is correct).

**Reply by xr0master** [reply]
It is quite clear in your screenshot that the public key (user ID) is not valid.

**Reply by danielbirch** [reply]
Ok so user_id is the public key? Good to know.
However the public key is 1 billion percent correct...
<img width="234" alt="Screenshot 2023-10-22 at 11 41 48 AM" src="https://github.com/emailjs-com/help/assets/122473191/c63c9d49-a2cf-4ba7-8956-a5c59392e48d">
<img width="1244" alt="Screenshot 2023-10-22 at 11 42 38 AM" src="https://github.com/emailjs-com/help/assets/122473191/70f96a4b-4530-45c4-a3a0-411a4fbdbbe8">

This is what's puzzling me.

**Reply by xr0master** [reply]
Most likely you override it somewhere else, I think you should debug the application. You can share with us the emailjs.sendForm code

**Reply by danielbirch** [reply]
I wish I did have it somewhere else, it would be an easy fix. My application is very simple, a portfolio and I checked again, the public key appears only one time. The only other area could be the order of the arguments for the .sendForm method?
<img width="605" alt="Screenshot 2023-10-22 at 12 15 16 PM" src="https://github.com/emailjs-com/help/assets/122473191/6090f941-a221-46d2-990e-7170479caac6">

I have been debugging it, when the form is submitted the error shows in the console and no events are logged in EmailJS. So that means the sendForm method is not triggering and looking at the error, it appears to be because the public key is invalid. So I expect...

A. The public key is incorrect - but it's not, I copied and pasted and you can see it's the same
B. The public key is in the wrong place - it's in the <head> right where the docs say it should be
C. There's another public key somewhere in the code - there isn't, I've checked my code several times
D. The public key may have expired? - which is why I regenerated a new public key just to be sure and still no luck

Could it be a scope issue, does the sendForm code need to be in the same code block as the emailjs.init function?

It's very frustrating, many months ago I had this working fine. Then EmailJs changed the placement of the public key in the code, so I updated my code to reflect these changes and I haven't been able to get it working since.

I appreciate your input on this @xr0master

**Reply by danielbirch** [reply]
Sure here is the code:

```
<script type="text/javascript" src="https://cdn.jsdelivr.net/npm/@emailjs/browser@3/dist/email.min.js"></script>
  
  <script type="text/javascript">
    (function() {
        // https://dashboard.emailjs.com/admin/account
        emailjs.init('ElvR8U91ELN6yUuSr');
    })();
  </script>

  <script type="text/javascript">
    document.addEventListener('DOMContentLoaded', function () {
// window.onload = function () {
	document
		.getElementById("contact-form")
		.addEventListener("submit", function (event) {
			console.log("Form submitted");
			event.preventDefault();
			// Generate five digit number for contact_number variable
			this.contact_number.value = (Math.random() * 100000) | 0;
			// // Grab recaptcha response
			var token = grecaptcha.getResponse();
			// Send and error handling
			emailjs
				.sendForm("contact_service", "contact_form",
					this,
					{ "g-recaptcha-response": token }
				)
				.then(
					function () {
						console.log("Success");
						document.getElementById("g-recaptcha").classList.add("hidden");
						const inputs = document.querySelectorAll("#user_name, #user_email, #message");
						inputs.forEach((input) => {
							input.value = "";
						});
					},
					function (error) {
						console.log("Failed", error);
					}
				);
		});
});
  </script>
```

**Comment by xr0master** [comment]
The 4th variable in the .sendForm function is the Public Key, but you put the object.
In your case, it will be
```js
emailjs.sendForm("contact_service", "contact_form", this)
```

**Reply by danielbirch** [reply]
If there 3rd parameter is the public key, where would this go then?

```
{ "g-recaptcha-response": token }
```

In the last version of this implementation, it was the last parameter.

**Reply by danielbirch** [reply]
If anyone knows the answer to this it would be massively helpful. 🙏🏻

**Reply by xr0master** [reply]
your code does not match the documentation, we showed you where the issue is. I don't even know what else you want to hear.

**Reply by danielbirch** [reply]
The documentation says nothing about where to place the recaptcha token capture code, only that you should include it. If you don't know either, that's fine mate. No need for hostility, you're not obliged to answer. I appreciate your assistance so far however, so thank you.

**Reply by xr0master** [reply, marked as answer]
The recaptcha code is taken automatically from the form, just fix your code and try to run :)

**Reply by danielbirch** [reply]
You sir are my new favourite person, thank you! It worked. 😊
